src/librustc/ty/query/plumbing.rs

```diff
@@ -601,9 +601,13 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
     pub(super) fn ensure_query<Q: QueryDescription<'gcx>>(self, key: Q::Key) -> () {
         let dep_node = Q::to_dep_node(self, &key);
 
-        // Ensuring an "input" or anonymous query makes no sense
+        if dep_node.kind.is_eval_always() {
```

Comment:
Ensuring queries just make sure the side effects of the query has been applied. This also applies to `eval_always` queries. However it doesn't make sense trying to mark them green with `try_mark_green_and_read` because we don't record dependencies for them, so they are handled separately here.